File: Assets/src/model/grasses/EveningBells.cs
using UnityEngine;

[System.Serializable]
[ObjectInfo(description: "Any non-player Creature walking into the Evening Bells falls into Deep Sleep for 3 turns. This consumes the Evening Bells.")]
public class EveningBells : Grass, IActorEnterHandler {
  public readonly float angle;

  public static bool CanOccupy(Tile tile) => tile is Ground && tile.CanBeOccupied();
  public EveningBells(Vector2Int pos, float angle) : base(pos) {
    this.angle = angle;
  }
  public EveningBells(Vector2Int pos) : this(pos, 0) {}

  public void HandleActorEnter(Actor actor) {
    if (actor != GameModel.main.player) {
      actor.SetTasks(new SleepTask(actor, 3, true));
      GameModel.main.EnqueueEvent(() => Kill(actor));
    }
  }
}

File: web/src/model/grasses/EveningBells.ts
import { Grass } from './Grass';
import { ACTOR_ENTER_HANDLER, type IActorEnterHandler } from '../../core/types';
import { Ground } from '../Tile';
import { GameModelRef } from '../GameModelRef';
import { SleepTask } from '../tasks/SleepTask';
import { entityRegistry } from '../../generator/entityRegistry';
import type { Vector2Int } from '../../core/Vector2Int';
import type { Tile } from '../Tile';

/**
 * Any non-player creature walking into the Evening Bells falls into
 * deep sleep for 3 turns. This consumes the Evening Bells.
 * Port of C# EveningBells.cs.
 */
export class EveningBells extends Grass implements IActorEnterHandler {
  readonly [ACTOR_ENTER_HANDLER] = true as const;
  readonly angle: number;

  constructor(pos: Vector2Int, angle = 0) {
    super(pos);
    this.angle = angle;
  }

  static canOccupy(tile: Tile): boolean {
    return tile instanceof Ground && tile.canBeOccupied();
  }

  handleActorEnter(actor: any): void {
    if (actor !== GameModelRef.main.player) {
      actor.setTasks(new SleepTask(actor, 3, true));
      GameModelRef.main.enqueuEvent(() => this.kill(actor));
    }
  }
}

entityRegistry.register('EveningBells', EveningBells);
